src/deadline/client/api/_session.py

```diff
@@ -67,36 +60,32 @@ def get_boto3_session(
     if profile_name in ("(default)", "default", ""):
         profile_name = None
 
-    # If a config was provided, don't use the Session caching mechanism.
-    if config:
-        return boto3.Session(profile_name=profile_name)
-
     if force_refresh:
         invalidate_boto3_session_cache()
 
-    # If this is the first call or the profile name has changed, make a fresh Session
-    if not __cached_boto3_session or __cached_boto3_session_profile_name != profile_name:
-        __cached_boto3_session = boto3.Session(profile_name=profile_name)
-        __cached_boto3_session_profile_name = profile_name
+    return _get_boto3_session_for_profile(profile_name)
 
-    return __cached_boto3_session
 
+@lru_cache
+def _get_boto3_session_for_profile(profile_name: str):
+    session = boto3.Session(profile_name=profile_name)
 
-def invalidate_boto3_session_cache() -> None:
-    """
-    Invalidates the cached boto3 session and boto3 queue session.
-    """
-    global __cached_boto3_session
-    global __cached_boto3_session_profile_name
-    global __cached_boto3_queue_session
-    global __cached_farm_id_for_queue_session
-    global __cached_queue_id_for_queue_session
+    # By default, DCM returns creds that expire after 15 minutes, and boto3's RefreshableCredentials
+    # class refreshes creds that are within 15 minutes of expiring, so credentials would never be reused.
+    # Also DCM credentials currently take several seconds to refresh. Set the refresh timeouts to 1 and
+    # 2 minutes so they are reused between API calls to save time.
+    # See https://github.com/boto/botocore/blob/develop/botocore/credentials.py#L342-L362
+    credentials = session.get_credentials()
+    if isinstance(credentials, RefreshableCredentials):
+        credentials._advisory_refresh_timeout = 2 * 60  # 2 minutes
+        credentials._mandatory_refresh_timeout = 60  # 1 minute
```

Comment:
This seems really aggressive to me. It doesn't leave much room for service issues before credentials expire; potentially during a long set of S3 uploads for job attachments. 

What would you think about advisory of 5 minutes and mandatory of 2.5? Still pretty aggressive, but leaves a little more leeway

Reply:
Sure, seems good to me. The creds' expiration is rechecked before every API call so I think anything over a few seconds is probably sufficient. I suspect the timeout/margin is to compensate for things like processing time between the creds check and API call or network delays. But leaving extra margin seems safe and won't noticably diminish responsiveness.